react/OrderForm.tsx

```diff
@@ -4,55 +4,139 @@ import React, {
+  orderForm: OrderForm
+  setOrderForm: (nextValue: OrderFormUpdate) => void
   error: ApolloError | undefined
 }
 
-const OrderFormContext = createContext<Context | undefined>(undefined)
+const noop = () => {}
+
+// keep default value as -1 to indicate this order form
+// is the initial value (not yet synchonized with server).
+const UNSYNC_ORDER_FORM_VALUE = -1
+
+const DEFAULT_ORDER_FORM: OrderForm = {
+  id: 'default-order-form',
+  items: [],
+  value: UNSYNC_ORDER_FORM_VALUE,
+  totalizers: [],
+  marketingData: {},
+  canEditData: false,
+  paymentData: {
+    installmentOptions: [],
+    paymentSystems: [],
+  },
+  messages: {
+    couponMessages: [],
+    generalMessages: [],
+  },
+  shipping: {},
+}
 
-const updateApolloCache = (client: ApolloClient<any>, orderForm: OrderForm) => {
-  const data = client.readQuery({ query: OrderFormQuery })
-  client.writeQuery({
-    query: OrderFormQuery,
-    data: {
-      ...data,
-      orderForm,
+const OrderFormContext = createContext<Context>({
+  orderForm: DEFAULT_ORDER_FORM,
+  setOrderForm: noop,
+  error: undefined,
+  loading: false,
+})
+
+const useUpdateOrderFormCache = () => {
+  const client = useApolloClient()
+
+  const updateOrderFormCache = useCallback(
+    (orderForm: OrderForm) => {
+      let data: Partial<OrderForm>
+
+      try {
+        data = client.readQuery({ query: OrderFormQuery }) ?? {}
+      } catch {
+        data = {}
+      }
+
+      client.writeQuery({
+        query: OrderFormQuery,
+        data: {
+          ...data,
+          orderForm,
+        },
+      })
     },
-  })
+    [client]
+  )
+
+  return updateOrderFormCache
+}
+
+const reducer = (
+  prevOrderForm: OrderForm,
+  updateOrderForm: OrderFormUpdate
+) => {
+  if (typeof updateOrderForm === 'function') {
+    return {
+      ...prevOrderForm,
+      ...updateOrderForm({
+        ...prevOrderForm,
+        value:
+          prevOrderForm.value === UNSYNC_ORDER_FORM_VALUE
+            ? 0
+            : prevOrderForm.value,
+      }),
+    }
+  }
+
+  return {
+    ...prevOrderForm,
+    ...updateOrderForm,
+  }
+}
+
+const saveLocalOrderForm = (orderForm: OrderForm) => {
+  localStorage.setItem('orderform', JSON.stringify(orderForm))
+}
+
+const getLocalOrderForm = (): OrderForm | null => {
+  let localOrderForm = null
+
+  try {
+    localOrderForm = JSON.parse(localStorage.getItem('orderform') ?? 'null')
+  } catch {
+    // ignore errors during SSR
+  }
+
+  return localOrderForm
 }
 
 export const OrderFormProvider: FC = ({ children }) => {
-  const { client, loading: loadingQuery, data, error } = useQuery<{
+  const { loading, data, error } = useQuery<{
     orderForm: OrderForm
   }>(OrderFormQuery, {
     ssr: false,
   })
 
   const [orderForm, setOrderForm] = useReducer(
-    (orderForm: OrderForm, newOrderForm: Partial<OrderForm>) => ({
-      ...orderForm,
-      ...newOrderForm,
-    }),
-    dummyOrderForm
+    reducer,
+    getLocalOrderForm() ?? DEFAULT_ORDER_FORM
```

Comment:
Oh, so this is the reason we don't need to update the Apollo cache anymore: you get the order form from the local storage, so the value in `data` from the query above (which is outdated) is ignored in subsequent renders.